src/mrpro/operators/WaveletOp.py

```diff
@@ -0,0 +1,353 @@
+"""Wavelet operator."""
+
+from collections.abc import Sequence
+
+import numpy as np
+import torch
+from ptwt.conv_transform import wavedec, waverec
+from ptwt.conv_transform_2 import wavedec2, waverec2
+from ptwt.conv_transform_3 import wavedec3, waverec3
+from pywt import Wavelet
+from pywt._multilevel import _check_level
+
+from mrpro.operators.LinearOperator import LinearOperator
+
+
+class WaveletOp(LinearOperator):
+    """Wavelet operator class."""
+
+    def __init__(
+        self,
+        domain_shape: Sequence[int] | None = None,
+        dim: tuple[int] | tuple[int, int] | tuple[int, int, int] = (-2, -1),
+        wavelet_name: str = 'db4',
+        level: int | None = None,
+    ):
+        """Wavelet operator.
+
+        For complex images the wavelet coefficients are calculated for real and imaginary part separately.
+        Wavelet names supported by pywt are: haar, db1 - db38, sym2 - sym20, coif1 - coif17, bior1.1, bior1.3, bior1.5,
+        bior2.2, bior2.4, bior2.6, bior2.8, bior3.1, bior3.3 bior3.5, bior3.7, bior3.9, bior4.4, bior5.5, bior6.8,
+        rbio1.1, rbio1.3, rbio1.5, rbio2.2, rbio2.4, rbio2.6, rbio2.8, rbio3.1, rbio3.3, rbio3.5, rbio3.7, rbio3.9,
+        rbio4.4, rbio5.5, rbio6.8,- bior6.8, dmey, gaus1- gaus8, mexh, cgau1 - cgau8, shan, fbsp, cmor
+
+        For a 2D image, the coefficients are labelled [aa, (ad_n, da_n, dd_n), ..., (ad_1, da_1, dd_1)] where a refers
+        to the approximation coefficients and d to the detail coefficients. The index indicates the level.
+
+        Parameters
+        ----------
+        domain_shape
+            Shape of domain where wavelets are calculated. If set to None the shape is taken from the input of the
+            forward operator. The adjoint operator will raise an error.
+        dim
+            Dimensions (axes) where wavelets are calculated
+        wavelet_name
+            Name of wavelets
+        level
+            Highest wavelet level. If set to None, the highest possible level is calculated based on the domain shape.
+
+        Raises
+        ------
+        ValueError
+            If wavelets are calculated for more than three dimensions.
+        ValueError
+            If wavelet dimensions and domain shape do not match.
+        NotImplementedError
+            If any dimension of the domain shape is odd. Adjoint will lead to the wrong domain shape.
+        """
+        super().__init__()
+        self._domain_shape = domain_shape
+        self._wavelet_name = wavelet_name
+        self._dim = dim
+        self._level = level
+
+        # number of wavelet directions
+        if len(dim) == 1:
```

Comment:
Use a dictionary to map dimensionality to 'n_wavelet_directions' instead of multiple if-else statements, which can simplify the logic and reduce code lines. [medium]